benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java

```diff
@@ -81,94 +89,147 @@
+        Benchmark firstBenchmark = benchmarks.get(0);
+        Map<Benchmark, BenchmarkExecutionResultBuilder> results = benchmarks.stream().collect(toMap(
+                Function.identity(),
+                benchmark -> new BenchmarkExecutionResultBuilder(benchmark).withExecutions(List.of())));
         List<QueryExecutionResult> executions;
         try {
-            executions = executeQueries(benchmark, benchmark.getPrewarmRuns(), true, executionTimeLimit);
-            String comparisonFailures = getComparisonFailuresString(executions);
-
+            executions = executeQueries(benchmarks, firstBenchmark.getPrewarmRuns(), true, executionTimeLimit);
+        }
+        catch (Exception e) {
+            return results.values().stream()
+                    .map(builder -> builder.withUnexpectedException(e).build())
+                    .collect(toList());
+        }
+        Map<Benchmark, String> comparisonFailures = getComparisonFailures(executions);
+        List<Benchmark> validBenchmarks = new ArrayList<>(benchmarks);
+        for (Map.Entry<Benchmark, BenchmarkExecutionResultBuilder> entry : results.entrySet()) {
+            Benchmark benchmark = entry.getKey();
+            BenchmarkExecutionResultBuilder result = entry.getValue();
             executionSynchronizer.awaitAfterBenchmarkExecutionAndBeforeResultReport(benchmark);
-
             statusReporter.reportBenchmarkStarted(benchmark);
-
-            resultBuilder = resultBuilder.startTimer();
-
-            try {
-                if (!comparisonFailures.isEmpty()) {
-                    throw new RuntimeException(format("Query result comparison failed for queries: %s", comparisonFailures));
-                }
-                executions = executeQueries(benchmark, benchmark.getRuns(), false, executionTimeLimit);
-                resultBuilder = resultBuilder.withExecutions(executions);
+            result.startTimer();
+            String failure = comparisonFailures.getOrDefault(benchmark, "");
+            if (!failure.isEmpty()) {
+                result.withUnexpectedException(new RuntimeException(format("Query result comparison failed for queries: %s", failure)));
+                result.endTimer();
+                validBenchmarks.remove(benchmark);
             }
-            finally {
-                resultBuilder = resultBuilder.endTimer();
-            }
-        }
-        catch (RuntimeException e) {
-            resultBuilder = resultBuilder.withUnexpectedException(e);
         }
 
-        BenchmarkExecutionResult executionResult = resultBuilder.build();
-
-        statusReporter.reportBenchmarkFinished(executionResult);
-
-        return executionResult;
+        try {
+            executions = executeQueries(validBenchmarks, firstBenchmark.getRuns(), false, executionTimeLimit);
+        }
+        catch (Exception e) {
+            return results.values().stream()
+                    .map(builder -> builder.withUnexpectedException(e).build())
```

Comment:
Would it be possible to assign the exception to a particular query/benchmark by catching the exception closer to the source? It would be easier to debug than just knowing "one of these benchmarks failed" .

Reply:
I just noticed `BenchmarkExecutionResult` has a reference to `Benchmark` so this isn't a big issue.

Reply:
We do this for query execution exceptions. This handles any other kind of issues, like concurrent execution. I'm not sure how to make it more fine-grained.